src/ui/public/kfetch/kfetch.ts

```diff
@@ -18,74 +18,91 @@
-    if (Error.captureStackTrace) {
-      Error.captureStackTrace(this, FetchError);
-    }
-  }
+interface KFetchQuery {
+  [key: string]: string | number | boolean;
 }
 
 export interface KFetchOptions extends RequestInit {
   pathname?: string;
-  query?: { [key: string]: string | number | boolean };
+  query?: KFetchQuery;
 }
 
 export interface KFetchKibanaOptions {
   prependBasePath?: boolean;
 }
 
-export function kfetch(fetchOptions: KFetchOptions, kibanaOptions?: KFetchKibanaOptions) {
-  // fetch specific options with defaults
-  const { pathname, query, ...combinedFetchOptions } = merge(
-    {
-      method: 'GET',
-      credentials: 'same-origin',
-      headers: {
-        'Content-Type': 'application/json',
-        'kbn-version': metadata.version,
-      },
-    },
-    fetchOptions
-  );
+export interface Interceptor {
+  request?: (config: any) => any;
+  requestError?: (e: any) => any;
+  response?: (res: any) => any;
+  responseError?: (e: any) => any;
+}
 
-  // kibana specific options with defaults
-  const combinedKibanaOptions = {
-    prependBasePath: true,
-    ...kibanaOptions,
-  };
+const interceptors: Interceptor[] = [];
+export const resetInterceptors = () => (interceptors.length = 0);
+export const addInterceptor = (interceptor: Interceptor) => interceptors.push(interceptor);
 
-  const fullUrl = url.format({
-    pathname: combinedKibanaOptions.prependBasePath ? chrome.addBasePath(pathname) : pathname,
-    query,
-  });
+export async function kfetch(
+  options: KFetchOptions,
+  { prependBasePath = true }: KFetchKibanaOptions = {}
+) {
+  const combinedOptions = {
+    method: 'GET',
+    credentials: 'same-origin',
+    ...options,
+    headers: {
+      'Content-Type': 'application/json',
+      'kbn-version': metadata.version,
+      ...options.headers,
+    },
+  };
 
-  const fetching = new Promise<any>(async (resolve, reject) => {
-    const res = await fetch(fullUrl, combinedFetchOptions);
+  const promise = requestInterceptors(combinedOptions).then(
+    ({ pathname, query, ...restOptions }: KFetchOptions) => {
+      const fullUrl = url.format({
+        pathname: prependBasePath ? chrome.addBasePath(pathname) : pathname,
+        query,
+      });
 
-    if (res.ok) {
-      return resolve(await res.json());
+      return fetch(fullUrl, restOptions).then(async res => {
+        if (res.ok) {
+          return res.json();
+        }
+        throw new KFetchError(res, await getBodyAsJson(res));
+      });
     }
+  );
 
-    try {
-      // attempt to read the body of the response
-      return reject(new FetchError(res, await res.json()));
-    } catch (_) {
-      // send FetchError without the body if we are not be able to read the body for some reason
-      return reject(new FetchError(res));
-    }
-  });
+  return responseInterceptors(promise);
+}
+
+function requestInterceptors(config: any) {
+  return [...interceptors].reverse().reduce((acc, interceptor) => {
```

Comment:
You could use `Array#reduceRight()` here and avoid the need for `.reverse()` which I think would mean you wouldn't have to spread-clone the interceptors array on each request.

Reply:
Maybe add a comment above this method explaining why we reverse/reduceRight here, and not in the response interceptors. I liked @spalger's explanation but I will 100% forget it within 2 weeks and wonder about this every time I ever look at it :P

Reply:
Oh of course. `reduceRight` to the rescue, thank you! And good idea with a comment to explain the behaviour!